### PR TITLE
feat: hashgraph

### PR DESCRIPTION
**Pull request by h3lio5**
adds a draft implementation of Hashgraph.

### CODE REVIEW

File: packages/object/src/hashgraph.ts [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+      }
+    }
+
+    // Initialize the queue
+    const queue: Hash[] = [];
+    outDegrees.forEach((key, value) => {
+      if (value === 0) {
+        queue.push(key);
+      }
+    }
```
Comment (d-roak) [inline packages/object/src/hashgraph.ts:124]: no conflict resolution here?

File: packages/object/src/hashgraph.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+    while (queue.length > 0) {
+      const nodeHash = queue.shift();
+      const nodeOp = this.nodes.get(nodeHash).operation;
+      result.push(nodeOp);
```
Comment (d-roak) [inline packages/object/src/hashgraph.ts:129]: huh? isn't the result the hash of the nodes?

File: packages/object/src/hashgraph.ts [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    [].concat(...this.dependencies)
+  }
+}
+
+class HashgraphDAG<O,T> {
+  private nodes: Map<Hash, HashgraphNode<T>> = new Map();
+  private frontier: Set<Hash> = new Set();
+  private originSet: Set<Hash> = new Set();
+  private precedenceRules: Map<string, boolean> = new Map();
+  private tickStore: Map<O, Hash> = new Map();
```
Comment (d-roak) [inline packages/object/src/hashgraph.ts:42]: seems the T and O are exchanged here

Comment (d-roak) [inline packages/object/src/hashgraph.ts:42]: shouldn't be Map<T, Set<Hash>> ?

**Comment by JanLewDev** [comment]
I found a paper about LSCA, which can be used in the future to improve the performance of the areCausallyRelated function. https://www.sciencedirect.com/science/article/pii/S0020019010000487

File: packages/crdt/src/builtins/AddWinsSet/index.ts
Comment (d-roak) [inline packages/crdt/src/builtins/AddWinsSet/index.ts:1]: as is, this is not a CRDT nor a CRO, it shouldn't exist in the builtins package. When I told you to separate it, the goal was to test the experience of creating a CRO that interacts with an HG. It should be a file like this:
- https://github.com/topology-foundation/ts-topology/blob/main/examples/canvas/src/objects/canvas.ts
- https://github.com/topology-foundation/ts-topology/blob/main/examples/chat/src/objects/chat.ts

Although this examples are currently being refactored

Comment (d-roak) [reply packages/crdt/src/builtins/AddWinsSet/index.ts:1]: basically, what u setted on the test file

File: packages/crdt/tests/AddWinsSet.test.ts [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+enum OperationType {
+    Add,
+    Remove
+}
+
+class Operation<T> {
+    constructor(
+        public readonly type: OperationType,
+        public readonly value: T
+    ) { }
```
Comment (d-roak) [inline packages/crdt/tests/AddWinsSet.test.ts:23]: not using??

File: packages/crdt/tests/AddWinsSet.test.ts [hunk 5]
```diff
@@ -0,0 +1,323 @@
+import { describe, test, expect, beforeEach } from "vitest";
+import { AddWinsSet } from "../src/builtins/AddWinsSet";
+import { ActionType, HashGraph } from "../../object/src/hashgraph";
+import { TopologyObject } from "../../object/src";
+
+interface IAddWinsCRO<T extends number> extends TopologyObject {
+    addWinsSet: AddWinsSet<T>;
+    hashGraph: HashGraph<Operation<T>>;
```
Comment (d-roak) [inline packages/crdt/tests/AddWinsSet.test.ts:8]: From our previous conversations, the HGs are not stored on the CRO, they are stored on the TopologyObject and have an interface to interact with it. This is not doing that

File: packages/crdt/tests/AddWinsSet.test.ts [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+        // Add the vertex V9 -> [V5]
+        let op9: Operation<number> = new Operation(OperationType.Remove, 1);
+        let deps9: string[] = [vertexHash5];
+        let vertexHash9 = cro.hashGraph.addVertex(op9, deps9);
+
+        let linearOps = cro.hashGraph.linearizeOps();
+        expect([[op1, op2, op6, op7, op4, op5]]).toContainEqual(linearOps);
+    });
+
+    test("Test: Jaoa's latest brain teaser", () => {
```
Comment (d-roak) [inline packages/crdt/tests/AddWinsSet.test.ts:284]:
```suggestion
    test("Test: Joao's latest brain teaser", () => {
```

File: packages/object/src/hashgraph.ts [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+  }
+
+  // Time complexity: O(V), Space complexity: O(V)
+  private areCausallyRelated(hash1: Hash, hash2: Hash): boolean {
+    const visited = new Set<Hash>();
+    const stack = [hash1];
+
+    while (stack.length > 0) {
+      const current = stack.pop()!;
+      if (current === hash2) return true;
```
Comment (d-roak) [inline packages/object/src/hashgraph.ts:145]: if `hash1 === hash2` it will return true, it shouldn't

File: packages/object/src/hashgraph.ts [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        }
+      }
+    }
+
+    visited.clear();
+    stack.push(hash2);
+
+    while (stack.length > 0) {
+      const current = stack.pop()!;
+      if (current === hash1) return true;
```
Comment (d-roak) [inline packages/object/src/hashgraph.ts:161]: same as above

File: packages/object/src/hashgraph.ts [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+      result.push(hash);
+
+      // Check if all dependencies are visited
+      if (Array.from(vertex.dependencies).every(dep => visited.has(dep))) {
+        const children = this.forwardEdges.get(hash) || new Set();
+        for (const child of children) {
+          // console.log("Visiting child: ", child);
+          visit(child);
+        }
+      }
```
Comment (d-roak) [inline packages/object/src/hashgraph.ts:81]: TODO: note to self to review tmw

File: packages/object/src/hashgraph.ts [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+    public readonly dependencies: Set<Hash>
+  ) { }
+}
+
+export enum ActionType {
+  DropLeft,
+  DropRight,
+  Nop,
+  Swap
+}
```
Comment (d-roak) [inline packages/object/src/hashgraph.ts:18]: It seems you are delegating the responsibility of defining the conflict resolution based on this types to the CRO devs. imo it's bad DX

File: packages/object/src/hashgraph.ts [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+      return hash; // Vertex already exists
+    }
+
+    const vertex = new Vertex(hash, operation, new Set(dependencies));
+    this.vertices.set(hash, vertex);
+    this.frontier.add(hash);
+
+    // update root
+    if (dependencies.length === 0) {
+      this.root = hash;
```
[the remark this answers is not in the document]
Comment (JanLewDev) [reply packages/object/src/hashgraph.ts:48]: What about creating a dummy, empty root that has no operation but every other op depends on it

File: packages/object/src/hashgraph.ts (hunk 10 above)
Comment (JanLewDev) [reply packages/object/src/hashgraph.ts:18]: Isn't conflict resolution defined here?
https://github.com/topology-foundation/ts-topology/pull/103/files#diff-c94ef3f2c7d03278c7a1c0d45f56ba471bca9cd7ecee78b183562aea23f9a8e6R37-R42
The action types are building blocks that can be used in different ways as far as I understand

File: packages/object/src/hashgraph.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (d-roak) [reply packages/object/src/hashgraph.ts:48]: Who creates it?

File: packages/object/src/hashgraph.ts (hunk 10 above)
Comment (d-roak) [reply packages/object/src/hashgraph.ts:18]: Exactly. In the CRO development. Does it look super straightforward for someone who might not understand crdt's and causal ordering?

File: packages/object/src/hashgraph.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (JanLewDev) [reply packages/object/src/hashgraph.ts:48]: The hashgraph itself, because it doesn't have an op

File: packages/crdt/tests/AddWinsSet.test.ts [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+        */
+
+        let op1: Operation<number> = new Operation(OperationType.Add, 1);
+        let deps1: string[] = [];
+        let vertexHash1 = cro.hashGraph.addVertex(op1, deps1);
+        // Add second vertex
+        let op2: Operation<number> = new Operation(OperationType.Add, 1);
+        let deps2: string[] = [vertexHash1];
+        let vertexHash2 = cro.hashGraph.addVertex(op2, deps2);
+        // Add the third vertex V3 concurrent with V2
```
Comment (JanLewDev) [inline packages/crdt/tests/AddWinsSet.test.ts:251]: // Add the third vertex V3 with dependency on V2

File: packages/crdt/tests/AddWinsSet.test.ts [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+        let op3: Operation<number> = new Operation(OperationType.Remove, 2);
+        let deps3: string[] = [vertexHash2];
+        let vertexHash3 = cro.hashGraph.addVertex(op3, deps3);
+        // Add the vertex V4 with dependency on V2
```
Comment (JanLewDev) [inline packages/crdt/tests/AddWinsSet.test.ts:255]: // Add the vertex V4 concurrent to V2

File: packages/crdt/package.json [hunk 14]
```diff
@@ -25,5 +25,7 @@
     "clean": "rm -rf dist/ node_modules/",
     "prepack": "tsc -b",
     "test": "vitest"
-  }
+  },
+  "dependencies": {},
+  "devDependencies": {}
```
Comment (d-roak) [inline packages/crdt/package.json:30]: if it's empty, you can just delete those

File: packages/crdt/src/builtins/AddWinsSet/index.ts
Comment (d-roak) [reply packages/crdt/src/builtins/AddWinsSet/index.ts:1]: not addressed

File: packages/crdt/src/builtins/AddWinsSet/index.ts [hunk 15]
```diff
@@ -0,0 +1,38 @@
+/// GSet with support for state and op changes
```
[the remark this answers is not in the document]
Comment (d-roak) [reply packages/crdt/src/builtins/AddWinsSet/index.ts:1]: not GSet

File: packages/crdt/tests/AddWinsSet.test.ts (hunk 5 above)
Comment (d-roak) [reply packages/crdt/tests/AddWinsSet.test.ts:8]: not addressed. What's the conclusion here? cc @guiltygyoza

File: packages/crdt/tests/AddWinsSet.test.ts (hunk 4 above)
Comment (d-roak) [reply packages/crdt/tests/AddWinsSet.test.ts:23]: not addressed

File: packages/object/package.json [hunk 16]
```diff
@@ -25,5 +25,7 @@
     "clean": "rm -rf dist/ node_modules/",
     "prepack": "tsc -b",
     "test": "vitest"
-  }
+  },
+  "dependencies": {},
+  "devDependencies": {}
```
Comment (d-roak) [inline packages/object/package.json:30]: if they are empty, you can rm this

File: packages/object/src/hashgraph.ts [hunk 17]
```diff
@@ -0,0 +1,199 @@
+import { sha256 } from "js-sha256";
+
+type Hash = string;
+
+class Vertex<T> {
+  constructor(
+    public readonly hash: Hash,
+    public readonly operation: T,
+    public readonly dependencies: Set<Hash>
```
[the remark this answers is not in the document]
Comment (d-roak) [reply packages/object/src/hashgraph.ts:9]: not addressed

File: packages/object/src/hashgraph.ts (hunk 10 above)
Comment (d-roak) [reply packages/object/src/hashgraph.ts:18]: decided to go with it for now

File: packages/object/src/hashgraph.ts [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+  DropRight,
+  Nop,
+  Swap
+}
+export class HashGraph<T> {
+  private vertices: Map<Hash, Vertex<T>> = new Map();
+  private frontier: Set<Hash> = new Set();
+  private forwardEdges: Map<Hash, Set<Hash>> = new Map();
+  // private resolveConflicts: (op1: T, op2: T) => ActionType;
+  private root: Hash = "";
```
[the remark this answers is not in the document]
Comment (d-roak) [reply packages/object/src/hashgraph.ts:24]: not addressed yet. If the idea is to add a NOP root, the hash is fixed

File: packages/object/src/hashgraph.ts (hunk 7 above)
Comment (d-roak) [reply packages/object/src/hashgraph.ts:176]: not addressed

File: packages/object/src/hashgraph.ts (hunk 8 above)
Comment (d-roak) [reply packages/object/src/hashgraph.ts:192]: not addressed

File: packages/object/tsconfig.json [hunk 19]
```diff
@@ -3,5 +3,5 @@
   "compilerOptions": {
     "outDir": "dist"
   },
-  "include": ["src/**/*.ts"]
+  "include": ["src/**/*.ts", "tests/hashgraph.test.ts"]
```
[the remark this answers is not in the document]
Comment (d-roak) [reply packages/object/tsconfig.json:6]: not addressed

**Review by d-roak — COMMENTED** [review]
there were some issues that were not addressed

**Pull request merged**